apollo-ios-codegen/Sources/IR/IR+RootFieldBuilder.swift

```diff
@@ -170,90 +163,147 @@ class RootFieldBuilder {
+
+        return condition
       }
-      let selectionSetScope = typeInfo.scope
+
+      let irFragmentSpread = buildNamedFragmentSpread(
+        fromFragment: fragmentSpread,
+        with: scope,
+        spreadIntoParentWithTypePath: typeInfo,
+        deferCondition: deferCondition
+      )
+      target.mergeIn(irFragmentSpread)
+
+    case (false, .some):
+      let irTypeCase = buildInlineFragmentSpread(
+        toWrap: selection,
+        with: scope,
+        inParentTypePath: typeInfo
+      )
+      target.mergeIn(irTypeCase)
+
+    case (false, nil):
+      let irTypeCaseEnclosingFragment = buildInlineFragmentSpread(
+        from: CompilationResult.SelectionSet(
+          parentType: fragmentSpread.parentType,
+          selections: [selection]
+        ),
+        with: scope,
+        inParentTypePath: typeInfo
+      )
+
+      target.mergeIn(irTypeCaseEnclosingFragment)
 
       var matchesType: Bool {
         guard let typeCondition = scope.type else { return true }
         return selectionSetScope.matches(typeCondition)
       }
-      let matchesScope = selectionSetScope.matches(scope)
 
-      if matchesScope {
-        let irFragmentSpread = buildNamedFragmentSpread(
-          fromFragment: fragmentSpread,
-          with: scope,
-          spreadIntoParentWithTypePath: typeInfo
-        )
-        target.mergeIn(irFragmentSpread)
-
-      } else {
-        let irTypeCaseEnclosingFragment = buildInlineFragmentSpread(
-          from: CompilationResult.SelectionSet(
-            parentType: fragmentSpread.parentType,
-            selections: [selection]
-          ),
-          with: scope,
-          inParentTypePath: typeInfo,
-          isDeferred: .init(fragmentSpread.isDeferred)
+      if matchesType {
+        typeInfo.entity.selectionTree.mergeIn(
+          selections: irTypeCaseEnclosingFragment.selectionSet.selections.direct.unsafelyUnwrapped.readOnlyView,
+          with: typeInfo
         )
-
-        target.mergeIn(irTypeCaseEnclosingFragment)
-
-        if matchesType {
-          typeInfo.entity.selectionTree.mergeIn(
-            selections: irTypeCaseEnclosingFragment.selectionSet.selections.direct.unsafelyUnwrapped.readOnlyView,
-            with: typeInfo
-          )
-        }
       }
     }
   }
 
   private func scopeCondition(
     for conditionalSelectionSet: ConditionallyIncludable,
-    in parentTypePath: SelectionSet.TypeInfo
+    in parentTypePath: SelectionSet.TypeInfo,
+    isDeferred: Bool = false
   ) -> ScopeCondition? {
     let inclusionResult = inclusionResult(for: conditionalSelectionSet.inclusionConditions)
     guard inclusionResult != .skipped else {
       return nil
     }
 
-    let type = parentTypePath.parentType == conditionalSelectionSet.parentType ?
-    nil : conditionalSelectionSet.parentType
+    let type = (
+      parentTypePath.parentType == conditionalSelectionSet.parentType
+      && !isDeferred
```

Comment:
Can we just add a comment here explaining _why_ we return `nil` when `isDeferred` is true? The logic is getting more and more complex and difficult to understand. These edge cases would be a lot easier to deal with in the future with inline explanations.

Reply:
Added in https://github.com/apollographql/apollo-ios-dev/pull/74/commits/ae4ce883f4e4819398e168d2008eb3ee1795dc66.